### PR TITLE
[Kernel] Use flashinfer for decoding

### PR DESCRIPTION
**Pull request by LiuXiaoxuanPKU**
This PR is a first attempt to integrate [flashinfer](https://flashinfer.ai/) for the decoding phase. The PR still uses flash attention for the prefill phase for now.

Updated after discussion with @yzh119
Things need to be fixed:
- [x] Use flashinfer's rope embedding --> Will turn off Flashinfer's rope support and use vllm's rope.
- [ ] Alibi slope --> Flashinfer will support alibi_slope as a input parameter.
- [ ] Cuda Graph support --> Flashinfer will simplify begin_forward function and add cuda graph support, no need to pad from the vllm side.

Next step:
- [x] Remove backend interface change
- [x] Add tests for e2e correctness
- [x] Add tests to check flashinfer integration works with tensor parallel

### CODE REVIEW

**Comment by LiuXiaoxuanPKU** [comment]
Pass the correctness test for single GPU and TP setting. Feel free to take a first pass @rkooo567

**Comment by rkooo567** [comment]
>  Cuda Graph support --> Flashinfer will simplify begin_forward function and add cuda graph support, no need to pad from the vllm side.

But we still need to pad inputs for other parts of kernel except attention?

File: csrc/cache_kernels.cu
Comment (rkooo567) [inline csrc/cache_kernels.cu:249]: need to add fp8 dtype like original rehape_and_cache? Otherwise, we should raise an exception when fp8 dtype is used for kv cache

File: tests/basic_correctness/test_flashinfer.py
Comment (rkooo567) [inline tests/basic_correctness/test_flashinfer.py:1]: Can we just use the same way as https://github.com/vllm-project/vllm/blob/4ea1f9678dd93f02424ab3de2149f83a490e6c6f/.buildkite/test-pipeline.yaml#L16 instead? You can do this in selector.py

File: tests/distributed/test_flashinfer_distributed.py
Comment (rkooo567) [inline tests/distributed/test_flashinfer_distributed.py:1]: same comment as above

File: vllm/utils.py
Comment (rkooo567) [inline vllm/utils.py:380]:
```suggestion
def create_kv_caches_with_random_flash(
```
(since it is same for both flash infer and attn)

File: vllm/attention/backends/abstract.py
Comment (rkooo567) [inline vllm/attention/backends/abstract.py:19]: thanks for catching this!

File: vllm/attention/backends/flashinfer.py
Comment (rkooo567) [inline vllm/attention/backends/flashinfer.py:109]: why is that? can you add a comment to code?

File: vllm/attention/backends/flashinfer.py
Comment (rkooo567) [inline vllm/attention/backends/flashinfer.py:39]: don't we need them?

File: vllm/attention/backends/flashinfer.py
Comment (rkooo567) [inline vllm/attention/backends/flashinfer.py:52]: this comment is not relevant after chunked prefill. Can you remove it?

File: vllm/attention/backends/flashinfer.py
Comment (rkooo567) [inline vllm/attention/backends/flashinfer.py:89]: consider `assert not self._is_prompt` since it is not allowed to use it for prefill now?

File: vllm/attention/backends/flashinfer.py
Comment (rkooo567) [inline vllm/attention/backends/flashinfer.py:124]: QQ: is sliding window supported already

File: vllm/attention/backends/flashinfer.py
Comment (rkooo567) [inline vllm/attention/backends/flashinfer.py:140]:
```
if attn_metadata.num_prefill_tokens > 0:
    assert attn_metadata.num_decode_tokens == 0, ("chunked prefill is not supported with flash infer yet")
if attn_metadata.num_decode_tokens > 0:
    attn_metadata.num_prefill_tokens == 0, ("chunked prefill is not supported with flash infer yet")
```

I can make this work as a follow up

File: vllm/attention/backends/flashinfer.py
Comment (rkooo567) [inline vllm/attention/backends/flashinfer.py:152]:
```suggestion
        if prefill_meta:= attn_metadata.prefill_metadata:
```

File: vllm/attention/backends/flashinfer.py
Comment (rkooo567) [inline vllm/attention/backends/flashinfer.py:153]: We should add a logic for prefix attention here (you can copy from other attention impl).

File: vllm/attention/backends/flashinfer.py
Comment (rkooo567) [inline vllm/attention/backends/flashinfer.py:170]: Isn't this already contiguous?

File: vllm/attention/backends/flashinfer.py
Comment (rkooo567) [inline vllm/attention/backends/flashinfer.py:113]: QQ: does flash infer support all attn head size?

File: vllm/worker/model_runner.py
Comment (rkooo567) [inline vllm/worker/model_runner.py:161]:
```suggestion
        self.flash_infer_workspace_buffer: <add type>
```

Also you don't need to set `= None` if you know it is not used unless it is lazily initialized. See https://github.com/vllm-project/vllm/blob/4ea1f9678dd93f02424ab3de2149f83a490e6c6f/vllm/worker/model_runner.py#L149

File: vllm/worker/model_runner.py
Comment (rkooo567) [inline vllm/worker/model_runner.py:446]:
```suggestion
        # The following fields are only for flash infer
        flash_paged_kv_indices: List[int] = []
```

Comment (rkooo567) [reply vllm/worker/model_runner.py:446]: Also can you update comment what each field means?

File: vllm/worker/model_runner.py
Comment (rkooo567) [inline vllm/worker/model_runner.py:490]: in this case, do we even need `paged_kv_indices`? can we just pass block_table directly?

File: vllm/worker/model_runner.py
Comment (rkooo567) [inline vllm/worker/model_runner.py:447]: comment why it has 0 already! (also the expected length is seq_group_num + 1)

File: vllm/worker/model_runner.py
Comment (rkooo567) [inline vllm/worker/model_runner.py:492]:
```suggestion
                last_page_len = seq_data.get_len() % self.block_size
```

File: vllm/worker/model_runner.py
Comment (rkooo567) [inline vllm/worker/model_runner.py:550]: QQ: what's the best practice buffer size here? can you add a comment in this code with a link?

File: vllm/attention/backends/flashinfer.py
Comment (rkooo567) [inline vllm/attention/backends/flashinfer.py:73]: Comment it is same as block_table in other attention?

Comment (rkooo567) [reply vllm/attention/backends/flashinfer.py:73]: also add a expected length

File: vllm/attention/backends/flashinfer.py
Comment (rkooo567) [inline vllm/attention/backends/flashinfer.py:70]: indptr is confusing. Can you add more specific definition?

File: vllm/attention/backends/flashinfer.py
Comment (rkooo567) [inline vllm/attention/backends/flashinfer.py:86]: Remove it? This is already in `AttentionMetadata` class

File: vllm/attention/backends/flashinfer.py
Comment (rkooo567) [inline vllm/attention/backends/flashinfer.py:84]: Remove it and just use `paged_kv_indices.shape[1]`?

File: vllm/attention/backends/flashinfer.py
Comment (rkooo567) [inline vllm/attention/backends/flashinfer.py:82]: Remove it and just use `query.shape[3]`?

File: vllm/attention/backends/flashinfer.py
Comment (rkooo567) [inline vllm/attention/backends/flashinfer.py:78]: remove and just get it from input?

**Review by rkooo567 — CHANGES_REQUESTED** [review]
Very clean!! many comments are nits. So it seems like

1. not working with prefill
2. not working with prefix caching
3. not working with chunked prefill

This is correct right? In this case, we should make sure we raise exception properly.

Also, for prefix attn kernel, does flash infer has equivalent?

File: tests/kernels/test_cache.py
Comment (jikunshang) [inline tests/kernels/test_cache.py:8]: can you refactor `reshape_and_cache_flash` into `_custom_ops.py` to make the interface unified?

**Comment by esmeetu** [comment]
@LiuXiaoxuanPKU Thank you for this PR! Can we support other prefill backend instead of just the flash attention? like XFormers.

File: vllm/worker/model_runner.py
Comment (esmeetu) [inline vllm/worker/model_runner.py:592]: This wrapper seems unnecessary to be passed.

**Comment by esmeetu** [comment]
> @LiuXiaoxuanPKU Thank you for this PR! Can we support other prefill backend instead of just the flash attention? like XFormers.

I have implemented a xformer version but the performance was not good(10%~20% throughout drop).

**Comment by rkooo567** [comment]
I think the goal is to use flash infer prefill eventually, so flash attn is replaced soon!

File: csrc/cache_kernels.cu
Comment (LiuXiaoxuanPKU) [reply csrc/cache_kernels.cu:249]: Will raise error for now, we can add fp8 support for the kv cache later.

File: tests/basic_correctness/test_flashinfer.py
Comment (LiuXiaoxuanPKU) [reply tests/basic_correctness/test_flashinfer.py:1]: The problem is that flashinfer does not support cudagraph yet (enforce_eager = False will fail the test), do we still want to share the test then?

File: vllm/attention/backends/flashinfer.py
Comment (LiuXiaoxuanPKU) [reply vllm/attention/backends/flashinfer.py:89]: When using flashinfer, we are also creating the FlashInferMetadata, which will also call `__post_init__` by default, here we want to skip the `__post_init__` if it's the prefill phase.

File: vllm/worker/model_runner.py
Comment (LiuXiaoxuanPKU) [reply vllm/worker/model_runner.py:490]: it's a bit subtle here, `paged_kv_indices` is 1D and does not have any padding, while `block_tables` i2D and might be padded.

File: tests/basic_correctness/test_flashinfer.py
Comment (LiuXiaoxuanPKU) [reply tests/basic_correctness/test_flashinfer.py:1]: Change test_basic_correctness, skip non-eager mode

File: vllm/attention/backends/flashinfer.py
Comment (LiuXiaoxuanPKU) [reply vllm/attention/backends/flashinfer.py:39]: we can defer it to the next pr

File: vllm/attention/backends/flashinfer.py
Comment (LiuXiaoxuanPKU) [reply vllm/attention/backends/flashinfer.py:166]: Lily: please raise exception here

File: vllm/attention/backends/flashinfer.py
Comment (LiuXiaoxuanPKU) [reply vllm/attention/backends/flashinfer.py:170]: Lily: benchmark performance of contiguous

File: vllm/attention/backends/flashinfer.py
Comment (LiuXiaoxuanPKU) [reply vllm/attention/backends/flashinfer.py:113]: No, only [64, 128, 256] are supported based on
https://github.com/flashinfer-ai/flashinfer/blob/3d55c71a62052c590c130897d3a3db49b14fcc34/include/flashinfer/utils.cuh#L157

**Comment by LiuXiaoxuanPKU** [comment]
> Very clean!! many comments are nits. So it seems like
> 
> 1. not working with prefill
> 2. not working with prefix caching
> 3. not working with chunked prefill
> 
> This is correct right? In this case, we should make sure we raise exception properly.
> 
> Also, for prefix attn kernel, does flash infer has equivalent?

1/2/3 Yes
 Also, for prefix attn kernel, does flash infer has equivalent? --> Yes, will add that in the next PR.

I fixed the comments, feel free to take another pass! @rkooo567

**Comment by LiuXiaoxuanPKU** [comment]
> @LiuXiaoxuanPKU Thank you for this PR! Can we support other prefill backend instead of just the flash attention? like XFormers.

Yeah will do! Plan to do that in a separate PR to limit the scope of current one. Feel free to review this PR, thanks!

**Review by rkooo567 — APPROVED** [review]

**Review by rkooo567 — APPROVED** [review]
LGTM!

**Comment by rkooo567** [comment]
there's an odd ci failure

**Comment by LiuXiaoxuanPKU** [comment]
> there's an odd ci failure
Seems some package/import error, will take a look tonight

**Comment by LiuXiaoxuanPKU** [comment]
> > there's an odd ci failure
> 
> Seems some package/import error, will take a look tonight

Caused by flashinfer support of pytorch 2.3, will update this pr once flashinfer adds the support.

**Comment by rkooo567** [comment]
@LiuXiaoxuanPKU what's the latest supported version from flash infer? Also, is this something we can just simply build it with torch 2.3?

**Comment by rkooo567** [comment]
Let's merge it without adding tests to CI? I think we don't need to be blocked by them. I will also create an issue to their repo for 2.3 support

**Comment by rkooo567** [comment]
https://github.com/flashinfer-ai/flashinfer/issues/227

File: vllm/attention/backends/flashinfer.py
Comment (ywang96) [inline vllm/attention/backends/flashinfer.py:67]:
```suggestion
    # be allocated/deacollated by the FlashInferMetadata object.
```

File: vllm/attention/backends/flashinfer.py
Comment (ywang96) [inline vllm/attention/backends/flashinfer.py:99]: Should we consider making `[64, 128, 256]` a constant variable?

File: vllm/attention/backends/flashinfer.py
Comment (ywang96) [reply vllm/attention/backends/flashinfer.py:102]: > When using flashinfer, we are also creating the FlashInferMetadata, which will also call __post_init__ by default, here we want to skip the __post_init__ if it's the prefill phase.

IMO it's worth putting a`NOTE` in the code regarding this explanation.

File: vllm/worker/model_runner.py
Comment (ywang96) [inline vllm/worker/model_runner.py:568]: Just curious -  is there any point of making workspace buffer size configurable?

File: vllm/attention/selector.py
Comment (ywang96) [inline vllm/attention/selector.py:50]: Shall we add a warning here that eager mode is enforced unless you plan to work on it in the next PR?

File: vllm/sequence.py
Comment (ywang96) [inline vllm/sequence.py:584]: Good catch!

**Review by ywang96 — COMMENTED** [review]
Left a few comments - Thanks for working on this!

File: vllm/worker/model_runner.py
Comment (LiuXiaoxuanPKU) [reply vllm/worker/model_runner.py:568]: Currently, it's always 16MB

**Review by cadedaniel — APPROVED** [review]
Approving to unblock @LiuXiaoxuanPKU given @rkooo567 's approval.

@ywang96 if you have any more comments provide them and @LiuXiaoxuanPKU can follow up in next PR

**Review by ywang96 — APPROVED** [review]
LGTM! I don't have any other comment ATM - please resolve the conflict though, thanks!

**Comment by rkooo567** [comment]
awseom! super excited to see performance with cuda graph!

**Comment by zhyncs** [comment]
Hi @LiuXiaoxuanPKU Is FlashInfer currently enabled by default? After testing the throughput on the ShareGPT dataset, there was no significant improvement on vLLM, and the gap with LMDeploy is still quite large (**8.40 vs 19.62**).
I'm not sure if there is a configuration error or something I haven't considered. Here is the reproduction method, please let me know if there are any mistakes. Thanks.


```bash
# env
NVIDIA A100-SXM4-80GB
PyTorch: 2.3.0+cu118
vLLM: 0.4.2+cu118
flash-attn: 2.5.8
LMDeploy: 0.4.0+81108ff

# vLLM Server
python3 -m vllm.entrypoints.openai.api_server --model /workdir/Meta-Llama-3-8B-Instruct
# vLLM Client
python3 benchmark_serving.py --backend vllm --dataset /workdir/ShareGPT_V3_unfiltered_cleaned_split.json --model /workdir/Meta-Llama-3-8B-Instruct

# LMDeploy Server
python3 -m lmdeploy serve api_server /workdir/Meta-Llama-3-8B-Instruct --cache-max-entry-count 0.95
# LMDeploy Client
python3 benchmark_serving.py --backend lmdeploy --dataset /workdir/ShareGPT_V3_unfiltered_cleaned_split.json --tokenizer /workdir/Meta-Llama-3-8B-Instruct --model llama3 --port 23333
```

```
# vLLM Result

============ Serving Benchmark Result ============
Successful requests:                     1000
Benchmark duration (s):                  119.11
Total input tokens:                      215196
Total generated tokens:                  186473
Request throughput (req/s):              8.40
Input token throughput (tok/s):          1806.76
Output token throughput (tok/s):         1565.61
---------------Time to First Token----------------
Mean TTFT (ms):                          32831.19
Median TTFT (ms):                        24628.54
P99 TTFT (ms):                           88550.36
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          154.28
Median TPOT (ms):                        149.54
P99 TPOT (ms):                           406.19
==================================================

# LMDeploy Result

============ Serving Benchmark Result ============
Successful requests:                     1000
Benchmark duration (s):                  50.98
Total input tokens:                      215196
Total generated tokens:                  187514
Request throughput (req/s):              19.62
Input token throughput (tok/s):          4221.44
Output token throughput (tok/s):         3678.41
---------------Time to First Token----------------
Mean TTFT (ms):                          18022.59
Median TTFT (ms):                        16949.77
P99 TTFT (ms):                           39935.77
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          30.41
Median TPOT (ms):                        28.74
P99 TPOT (ms):                           77.57
==================================================
```

**Comment by LiuXiaoxuanPKU** [comment]
Hi @zhyncs, thanks for the interest and benchmarking, several things here:
FlashInfer is not turned on by default, it can only be enabled with environment variable `VLLM_ATTENTION_BACKEND=FLASHINFER`.

We don't turn it on by default because of performance concerns:
1. We don't have cuda graph support for flashinfer yet. Without cudagraph, it might be hard to see any performance benefits, though we have not thoroughly benchmarked it yet.
2. Flashinfer's begin_forward function will cause some extra CPU/GPU communication, which will be optimized from the flashinfer side.

1/2 are both doable and fixable. We will coordinate with Flashinfer team. After fixing the performance issue, we can turn Flashinfer on by default.

**Comment by zhyncs** [comment]
> Hi @zhyncs, thanks for the interest and benchmarking, several things here:
> 
> FlashInfer is not turned on by default, it can only be enabled with environment variable `VLLM_ATTENTION_BACKEND=FLASHINFER`.
> 
> 
> 
> We don't turn it on by default because of performance concerns:
> 
> 1. We don't have cuda graph support for flashinfer yet. Without cudagraph, it might be hard to see any performance benefits, though we have not thoroughly benchmarked it yet.
> 
> 2. Flashinfer's begin_forward function will cause some extra CPU/GPU communication, which will be optimized from the flashinfer side.
> 
> 
> 
> 1/2 are both doable and fixable. We will coordinate with Flashinfer team. After fixing the performance issue, we can turn Flashinfer on by default.
> 
> 

Thanks for your reply!

**Comment by MichoChan** [comment]
> Hi @zhyncs, thanks for the interest and benchmarking, several things here: FlashInfer is not turned on by default, it can only be enabled with environment variable `VLLM_ATTENTION_BACKEND=FLASHINFER`.
> 
> We don't turn it on by default because of performance concerns:
> 
> 1. We don't have cuda graph support for flashinfer yet. Without cudagraph, it might be hard to see any performance benefits, though we have not thoroughly benchmarked it yet.
> 2. Flashinfer's begin_forward function will cause some extra CPU/GPU communication, which will be optimized from the flashinfer side.
> 
> 1/2 are both doable and fixable. We will coordinate with Flashinfer team. After fixing the performance issue, we can turn Flashinfer on by default.

hi, when turn on by default？

**Comment by MichoChan** [comment]
> > Hi @zhyncs, thanks for the interest and benchmarking, several things here: FlashInfer is not turned on by default, it can only be enabled with environment variable `VLLM_ATTENTION_BACKEND=FLASHINFER`.
> > We don't turn it on by default because of performance concerns:
> > 
> > 1. We don't have cuda graph support for flashinfer yet. Without cudagraph, it might be hard to see any performance benefits, though we have not thoroughly benchmarked it yet.
> > 2. Flashinfer's begin_forward function will cause some extra CPU/GPU communication, which will be optimized from the flashinfer side.
> > 
> > 1/2 are both doable and fixable. We will coordinate with Flashinfer team. After fixing the performance issue, we can turn Flashinfer on by default.
> 
> hi, when turn on by default？

i found when qps or seq len is not large , the speed is slower than vllm base decode kernel, may 2 is needed to remove cpu/gpu copy

**Comment by rkooo567** [comment]
@MichoChan I think it is because it doesn't have cuda graph support yet (at large qps, the cpu overhead cuda graph removes is negligible usually)

**Comment by Calculusss** [comment]
> Hi @zhyncs, thanks for the interest and benchmarking, several things here: FlashInfer is not turned on by default, it can only be enabled with environment variable `VLLM_ATTENTION_BACKEND=FLASHINFER`.
> 
> We don't turn it on by default because of performance concerns:
> 
> 1. We don't have cuda graph support for flashinfer yet. Without cudagraph, it might be hard to see any performance benefits, though we have not thoroughly benchmarked it yet.
> 2. Flashinfer's begin_forward function will cause some extra CPU/GPU communication, which will be optimized from the flashinfer side.
> 
> 1/2 are both doable and fixable. We will coordinate with Flashinfer team. After fixing the performance issue, we can turn Flashinfer on by default.

Hi @LiuXiaoxuanPKU, Do you have a timeline regarding the support for CUDA Graphs?

**Comment by rkooo567** [comment]
We are blocked by flash infer side to enable cuda graph support. Should be able to support after that (I assume 2+weeks to get it delivered )

**Pull request merged**
